File: src/domain/timeline.ts
import { LIMITS } from './limits';
import { minutesBetween } from './time';
import type { ActivityEntry, ActivityType } from './types';

/** Непрерывный отрезок одного режима: соседние записи одного типа склеены. */
export interface Block {
  activity: ActivityType;
  start: number;
  /** Для текущего режима — момент расчёта. */
  end: number;
  open: boolean;
  entryIds: string[];
  /** Все записи блока сделаны в режиме «паром / поезд». */
  ferry: boolean;
  /** Водитель завершил этим отдыхом рабочий день. */
  dayEnd: boolean;
}

export const blockMinutes = (b: Block): number => minutesBetween(b.start, b.end);

// Записи, созданные переключением режима, стыкуются точно; допуск — на округление.
const CONTIGUITY_MS = 1000;

/**
 * Приводит журнал к последовательности блоков: сортирует, обрезает
 * наложения (приоритет у более ранней записи) и склеивает соседние записи
 * одного режима. Повторное нажатие на тот же режим или ручная правка не
 * должны дробить перерыв: 20 + 25 мин отдыха подряд — это 45 мин.
 */
export function buildBlocks(entries: ActivityEntry[], now: number): Block[] {
  const sorted = [...entries].sort((a, b) => a.startTime - b.startTime);
  const blocks: Block[] = [];

  for (const entry of sorted) {
    const open = entry.endTime === null;
    const end = Math.min(entry.endTime ?? now, now);
    const prev = blocks[blocks.length - 1];
    const start = prev ? Math.max(entry.startTime, Math.min(prev.end, end)) : entry.startTime;
    if (end <= start && !open) continue;

    if (prev && prev.activity === entry.activity && start - prev.end <= CONTIGUITY_MS) {
      prev.end = Math.max(prev.end, end);
      prev.open = prev.open || open;
      prev.entryIds.push(entry.id);
      prev.ferry = prev.ferry && !!entry.ferry;
      prev.dayEnd = prev.dayEnd || !!entry.dayEnd;
      continue;
    }
    blocks.push({
      activity: entry.activity,
      start,
      end: Math.max(start, end),
      open,
      entryIds: [entry.id],
      ferry: !!entry.ferry,
      dayEnd: !!entry.dayEnd,
    });
  }
  return blocks;
}

/** Период отдыха: один блок или несколько, прерванных на пароме / поезде. */
export interface RestPeriod {
  start: number;
  end: number;
  /** Чистое время отдыха без прерываний. */
  restMinutes: number;
  open: boolean;
  /** Отдых объявлен концом рабочего дня. */
  dayEnd: boolean;
  firstBlock: number;
  lastBlock: number;
}

/**
 * Находит периоды отдыха. Блоки отдыха объединяются через прерывания на
 * пароме / поезде (ст. 9: не больше двух, суммарно до 1 ч), только если
 * в сумме получается не меньше полного суточного отдыха (11 ч) — иначе
 * исключение не применяется.
 */
export function findRestPeriods(blocks: Block[]): RestPeriod[] {
  const periods: RestPeriod[] = [];
  let i = 0;
  while (i < blocks.length) {
    if (blocks[i].activity !== 'REST') {
      i++;
      continue;
    }
    const single = periodOf(blocks, i, i);
    let best = single;
    let interruptions = 0;
    let interruptionMinutes = 0;
    let last = i;

    while (true) {
      let j = last + 1;
      let runMinutes = 0;
      let ferryRun = true;
      let contiguous = true;
      while (j < blocks.length && blocks[j].activity !== 'REST') {
        ferryRun = ferryRun && blocks[j].ferry;
        contiguous = contiguous && blocks[j].start - blocks[j - 1].end <= CONTIGUITY_MS;
        runMinutes += blockMinutes(blocks[j]);
        j++;
      }
      if (j >= blocks.length || j === last + 1 || !ferryRun || !contiguous) break;
      if (blocks[j].start - blocks[j - 1].end > CONTIGUITY_MS) break;
      interruptions++;
      interruptionMinutes += runMinutes;
      if (
        interruptions > LIMITS.ferryMaxInterruptions ||
        interruptionMinutes > LIMITS.ferryMaxInterruptionMinutes
      ) {
        break;
      }
      last = j;
      const merged = periodOf(blocks, i, last);
      if (merged.restMinutes >= LIMITS.dailyRestRegular) best = merged;
    }

    periods.push(best);
    i = best.lastBlock + 1;
  }
  return periods;
}

function periodOf(blocks: Block[], first: number, last: number): RestPeriod {
  let restMinutes = 0;
  let dayEnd = false;
  for (let k = first; k <= last; k++) {
    if (blocks[k].activity !== 'REST') continue;
    restMinutes += blockMinutes(blocks[k]);
    dayEnd = dayEnd || blocks[k].dayEnd;
  }
  return {
    start: blocks[first].start,
    end: blocks[last].end,
    restMinutes,
    open: blocks[last].open,
    dayEnd,
    firstBlock: first,
    lastBlock: last,
  };
}

File: packages/tacho_engine/lib/src/timeline.dart
import 'package:meta/meta.dart';
import 'package:tacho_engine/src/activity_period.dart';
import 'package:tacho_engine/src/driver_mode.dart';
import 'package:tacho_engine/src/eu_limits.dart';
import 'package:tacho_engine/src/time.dart';

/// Непрерывный отрезок одного режима: соседние записи одного режима склеены.
@immutable
class Block {
  const new({
    required this.mode,
    required this.start,
    required this.end,
    required this.open,
    required this.ferry,
    required this.dayEnd,
  });

  final DriverMode mode;
  final DateTime start;

  /// Для текущего режима — момент расчёта.
  final DateTime end;
  final bool open;

  /// Все записи блока сделаны в режиме «паром / поезд».
  final bool ferry;

  /// Водитель завершил этим отдыхом рабочий день.
  final bool dayEnd;

  Duration get duration => durationBetween(start, end);

  @override
  String toString() => 'Block($mode, $start → $end${open ? ', open' : ''})';
}

/// Записи, созданные переключением режима, стыкуются точно; допуск — на
/// округление.
const contiguityTolerance = Duration(seconds: 1);

bool _contiguous(DateTime previousEnd, DateTime nextStart) =>
    nextStart.difference(previousEnd) <= contiguityTolerance;

/// Приводит журнал к последовательности блоков: сортирует, обрезает
/// наложения (приоритет у более ранней записи) и склеивает соседние записи
/// одного режима. Повторное нажатие на тот же режим или ручная правка не
/// должны дробить перерыв: 20 + 25 мин отдыха подряд — это 45 мин.
List<Block> buildBlocks(Iterable<ActivityPeriod> periods, DateTime now) {
  final sorted = sortedByStart(periods);
  final blocks = <Block>[];

  for (final p in sorted) {
    final open = p.isOpen;
    final end = earlier(p.end ?? now, now);
    final prev = blocks.isEmpty ? null : blocks.last;
    final start = prev == null
        ? p.start
        : later(p.start, earlier(prev.end, end));
    if (!end.isAfter(start) && !open) continue;

    if (prev != null && prev.mode == p.mode && _contiguous(prev.end, start)) {
      blocks.last = Block(
        mode: prev.mode,
        start: prev.start,
        end: later(prev.end, end),
        open: prev.open || open,
        ferry: prev.ferry && p.ferry,
        dayEnd: prev.dayEnd || p.dayEnd,
      );
      continue;
    }
    blocks.add(
      Block(
        mode: p.mode,
        start: start,
        end: later(start, end),
        open: open,
        ferry: p.ferry,
        dayEnd: p.dayEnd,
      ),
    );
  }
  return blocks;
}

/// Записи по возрастанию начала. Сортировка устойчивая: при равном начале
/// сохраняется порядок хранилища, как в прототипе.
List<ActivityPeriod> sortedByStart(Iterable<ActivityPeriod> periods) {
  final indexed = periods.indexed.toList()
    ..sort((a, b) {
      final byStart = a.$2.start.compareTo(b.$2.start);
      return byStart != 0 ? byStart : a.$1.compareTo(b.$1);
    });
  return [for (final (_, p) in indexed) p];
}

/// Период отдыха: один блок или несколько, прерванных на пароме / поезде.
@immutable
class RestPeriod {
  const new({
    required this.start,
    required this.end,
    required this.rest,
    required this.open,
    required this.dayEnd,
    required this.firstBlock,
    required this.lastBlock,
  });

  final DateTime start;
  final DateTime end;

  /// Чистое время отдыха без прерываний.
  final Duration rest;
  final bool open;

  /// Отдых объявлен концом рабочего дня.
  final bool dayEnd;

  /// Индексы первого и последнего блока периода в шкале.
  final int firstBlock;
  final int lastBlock;

  /// Отдых, который завершает смену: суточный (≥ 9 ч) или недельный, а также
  /// идущий отдых, которым водитель завершил день.
  bool get endsShift => rest >= EuLimits.dailyRestReduced || (open && dayEnd);

  /// Недельный отдых: не меньше сокращённого (24 ч).
  bool get isWeekly => rest >= EuLimits.weeklyRestReduced;

  @override
  String toString() =>
      'RestPeriod($start → $end, ${rest.inMinutes} мин'
      '${open ? ', open' : ''})';
}

/// Находит периоды отдыха. Блоки отдыха объединяются через прерывания на
/// пароме / поезде (ст. 9: не больше двух, суммарно до 1 ч), только если
/// в сумме получается не меньше полного суточного отдыха (11 ч) — иначе
/// исключение не применяется.
List<RestPeriod> findRestPeriods(List<Block> blocks) {
  final periods = <RestPeriod>[];
  var i = 0;
  while (i < blocks.length) {
    if (!blocks[i].mode.isRest) {
      i++;
      continue;
    }
    var best = _periodOf(blocks, i, i);
    var interruptions = 0;
    var interruptionTotal = Duration.zero;
    var last = i;

    while (true) {
      var j = last + 1;
      var run = Duration.zero;
      var ferryRun = true;
      var contiguous = true;
      while (j < blocks.length && !blocks[j].mode.isRest) {
        ferryRun = ferryRun && blocks[j].ferry;
        contiguous =
            contiguous && _contiguous(blocks[j - 1].end, blocks[j].start);
        run += blocks[j].duration;
        j++;
      }
      if (j >= blocks.length || j == last + 1 || !ferryRun || !contiguous) {
        break;
      }
      if (!_contiguous(blocks[j - 1].end, blocks[j].start)) break;
      interruptions++;
      interruptionTotal += run;
      if (interruptions > EuLimits.ferryMaxInterruptions ||
          interruptionTotal > EuLimits.ferryMaxInterruptionTotal) {
        break;
      }
      last = j;
      final merged = _periodOf(blocks, i, last);
      if (merged.rest >= EuLimits.dailyRestRegular) best = merged;
    }

    periods.add(best);
    i = best.lastBlock + 1;
  }
  return periods;
}

RestPeriod _periodOf(List<Block> blocks, int first, int last) {
  var rest = Duration.zero;
  var dayEnd = false;
  for (var k = first; k <= last; k++) {
    if (!blocks[k].mode.isRest) continue;
    rest += blocks[k].duration;
    dayEnd = dayEnd || blocks[k].dayEnd;
  }
  return RestPeriod(
    start: blocks[first].start,
    end: blocks[last].end,
    rest: rest,
    open: blocks[last].open,
    dayEnd: dayEnd,
    firstBlock: first,
    lastBlock: last,
  );
}
